### PR TITLE
Adding support for mamba

### PR DESCRIPTION
**Pull request by Goekdeniz-Guelmez**
Support for Mamba v1

### CODE REVIEW

**Comment by Goekdeniz-Guelmez** [comment]
Is there a way to add a Depthwise Conv1d directly instead of a class?

**Comment by awni** [comment]
>  Is there a way to add a Depthwise Conv1d directly instead of a class?

You can use the op [`mx.conv1d`](https://ml-explore.github.io/mlx/build/html/python/_autosummary/mlx.core.conv1d.html#mlx.core.conv1d) with the `groups` parameter.

File: llms/mlx_lm/models/mamba.py [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+        y = y + D * x
+        return y, h
+    
+    def __call__(self, x, cache):
+        h, inputs = cache
+        
+        x, z = self.in_proj(x).split(indices_or_sections=2, axis=-1)
+        
+        # x is now (B, L, C), we need (B, C, L) for conv1d
+        x_cache = x.transpose(0, 2, 1)
```
Comment (awni) [inline llms/mlx_lm/models/mamba.py:185]: Convolutions in MLX are channels last, so you shouldn't need to transpose anything (just make sure to transpose the weights when you convert the model, kind of like [here](https://github.com/ml-explore/mlx-examples/blob/main/llms/mlx_lm/models/recurrent_gemma.py#L510-L517))

**Comment by Goekdeniz-Guelmez** [comment]
so generating is almost done but the output Logits are floats and not ints:

```
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/Users/gokdenizgulmez/Desktop/mlx-examples-drafts/llms/mlx_lm/generate.py", line 243, in <module>
    main()
  File "/Users/gokdenizgulmez/Desktop/mlx-examples-drafts/llms/mlx_lm/generate.py", line 226, in main
    response = generate(
               ^^^^^^^^^
  File "/Users/gokdenizgulmez/Desktop/mlx-examples-drafts/llms/mlx_lm/utils.py", line 353, in generate
    for (token, logprobs), n in zip(
  File "/Users/gokdenizgulmez/Desktop/mlx-examples-drafts/llms/mlx_lm/utils.py", line 260, in generate_step
    y, logprobs = _step(y)
                  ^^^^^^^^
  File "/Users/gokdenizgulmez/Desktop/mlx-examples-drafts/llms/mlx_lm/utils.py", line 240, in _step
    logits = logits[:, -1, :]
             ~~~~~~^^^^^^^^^^
TypeError: tuple indices must be integers or slices, not tuple
```

The shapes are:
`Logits shape: (1, 5, 50280) --> (B, L, vocab_size)`

When printing them out i get:
```python
array([[[12.1114, -25.3092, 0.578034, ..., -25.1183, -24.7811, -24.9419],
        [16.8882, -44.0154, 15.631, ..., -43.858, -43.3342, -43.5788],
        [17.0612, -44.6536, 15.7307, ..., -44.486, -43.9664, -44.2006],
        [10.1423, 6.21938, 5.05909, ..., 6.02884, 5.97486, 6.19409],
        [1.13879, -3.57978, -1.9936, ..., -3.5286, -3.41974, -3.51589]]], dtype=float32)
```

**Comment by Goekdeniz-Guelmez** [comment]
Generating works but it's random stuff, probably something wrong with the cache handling.

**Comment by Goekdeniz-Guelmez** [comment]
Tried it with:

```sh
python -m mlx_lm.generate \                                                                                                                                                                                                                                                                                                            (adding-support-for-mamba|✚4)
    --model /Users/gokdenizgulmez/Library/Mobile\ Documents/com\~apple\~CloudDocs/Transformer\ Models/Safetensors/mamba-370m-hf \
    --prompt "hello how are you." \
    --colorize
```

**Comment by awni** [comment]
Great work @Goekdeniz-Guelmez. Let me know once you've tested training and we can get this landed!

**Comment by Goekdeniz-Guelmez** [comment]
> Great work @Goekdeniz-Guelmez. Let me know once you've tested training and we can get this landed!

Thanks. Will do!

**Comment by Goekdeniz-Guelmez** [comment]
Hey awni,

when trying to train it, I get `ValueError: [grad] Must specify at least one argument.`. upon looking at the trainable parameters for LoRA `Trainable parameters: 0.000% (0.000M/128.509M)` so the LoRA doesn't get applied to it correctly, even though I selected all the LoRA layers in the terminal and in the `trainer/utils` code `num_layers = len(model.backbone.layers)`.

Her is the full log:

```text
(mlx) (base)  ~/Desktop/mlx-examples-drafts/llms/ [adding-support-for-mamba] python -m mlx_lm.lora \                                                                (adding-support-for-mamba|✔)
    --model state-spaces/mamba-130m-hf \
    --train \
    --data /Users/gokdenizgulmez/Library/Mobile\ Documents/com\~apple\~CloudDocs/Datastes/data_tyni \
    --iters 100 \
    --batch-size 1 \
    --lora-layers 24 \
    --val-batches 1 \
    --adapter-path /Users/gokdenizgulmez/Desktop/mamba-pretrain
Loading pretrained model
Fetching 5 files: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:00<00:00, 50533.78it/s]
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
Loading datasets
Training
Trainable parameters: 0.000% (0.000M/128.509M)
Starting training..., iters: 100
Iter 1: Val loss 7.719, Val took 16.461s
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/Users/gokdenizgulmez/Desktop/mlx-examples-drafts/llms/mlx_lm/lora.py", line 278, in <module>
    main()
  File "/Users/gokdenizgulmez/Desktop/mlx-examples-drafts/llms/mlx_lm/lora.py", line 274, in main
    run(types.SimpleNamespace(**args))
  File "/Users/gokdenizgulmez/Desktop/mlx-examples-drafts/llms/mlx_lm/lora.py", line 247, in run
    train_model(args, model, tokenizer, train_set, valid_set, training_callback)
  File "/Users/gokdenizgulmez/Desktop/mlx-examples-drafts/llms/mlx_lm/lora.py", line 203, in train_model
    train(
  File "/Users/gokdenizgulmez/Desktop/mlx-examples-drafts/llms/mlx_lm/tuner/trainer.py", line 246, in train
    lvalue, toks = step(batch)
                   ^^^^^^^^^^^
  File "/Users/gokdenizgulmez/Desktop/mlx-examples-drafts/llms/mlx_lm/tuner/trainer.py", line 193, in step
    (lvalue, toks), grad = loss_value_and_grad(model, *batch)
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/miniconda3/envs/mlx/lib/python3.12/site-packages/mlx/nn/utils.py", line 34, in wrapped_value_grad_fn
    value, grad = value_grad_fn(model.trainable_parameters(), *args, **kwargs)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: [grad] Must specify at least one argument.
```

**Comment by Goekdeniz-Guelmez** [comment]
Fixed!! Training works now.

tried with:

```shell
python -m mlx_lm.lora \                                                                                                                                                                                                                                                                                                                            (adding-support-for-mamba|✔)
    --model state-spaces/mamba-130m-hf \
    --train \
    --data /Users/gokdenizgulmez/Library/Mobile\ Documents/com\~apple\~CloudDocs/Datastes/data_tyni \
    --iters 2 \ 
    --batch-size 1 \
    --lora-layers 2 \
    --val-batches 1 \
    --steps-per-report 2 \
    --adapter-path /Users/gokdenizgulmez/Desktop/mamba-pretrain
```

Output:

```text
Loading pretrained model
Fetching 5 files: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:00<00:00, 42027.09it/s]
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
Loading datasets
Training
Trainable parameters: 0.116% (0.150M/128.509M)
Starting training..., iters: 2
Iter 1: Val loss 7.719, Val took 15.770s
Iter 2: Val loss 7.734, Val took 13.569s
Iter 2: Train loss 7.451, Learning Rate 1.000e-05, It/sec 0.066, Tokens/sec 105.414, Trained Tokens 3200, Peak mem 2.955 GB
```

---

Generating with the LoRA adapters:

```shell
python -m mlx_lm.generate --model state-spaces/mamba-130m-hf --adapter-path /Users/gokdenizgulmez/Desktop/mamba-pretrain --prompt "Gökdeniz: hello how are you. JOSIE: What's up. Gökdeniz:"
```

Output:

```text
Fetching 5 files: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:00<00:00, 60963.72it/s]
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
==========
Prompt: Gökdeniz: hello how are you. JOSIE: What's up. Gökdeniz:
 You're welcome. I'll see you tomorrow. JOSIE: But I'm glad to see you. Gökdeniz: I see you too. JOSIE: All right. I see you too. Gökdeniz: I see you too. JOSIE: All right. I see you too. Gökdeniz: I see you too. JOSIE: All right. I see you too. Gökdeniz: I see you too. JOS
==========
Prompt: 23 tokens, 54.660 tokens-per-sec
Generation: 100 tokens, 65.461 tokens-per-sec
Peak memory: 0.490 GB
```

**Comment by Goekdeniz-Guelmez** [comment]
However, I can't try MistralAIs Mamba model because I got a M1 8GB

**Comment by Goekdeniz-Guelmez** [comment]
Here is the [Huggingface repo](https://huggingface.co/mlx-community/mamba-130m-hf-float32).

**Review by awni — APPROVED** [review]
Thanks for the awesome addition!!

**Comment by Goekdeniz-Guelmez** [comment]
No Problem!!! Version 2 will is also in the works.

**Comment by arogister** [comment]
I just tried generation with https://huggingface.co/mlx-community/mamba-2.8b-hf-f32 and noticed that mamba support was not in latest mlx-lm (0.18.2). It works fine when building from source though.

**Comment by awni** [comment]
Indeed we haven't done a release with it yet. It will be out in 19.0 shortly.

**Pull request merged**